src/Mod/Sketcher/Gui/TaskSketcherElements.cpp

```diff
@@ -677,128 +955,78 @@ void TaskSketcherElements::leaveEvent (QEvent * event)
-                                                                (QString::fromLatin1("%1-").arg(i)+tr("Hyperbolic Arc")))   :
-        type == Part::GeomArcOfParabola::getClassTypeId()    ? ( isNamingBoxChecked ?
-                                                                (tr("Parabolic Arc") + QString::fromLatin1("(Edge%1#ID%2)").arg(i).arg(i-1)) +
-                                                                (construction?(QString::fromLatin1("-")+tr("Construction")):QString::fromLatin1("")):
-                                                                (QString::fromLatin1("%1-").arg(i)+tr("Parabolic Arc")))   :
-        type == Part::GeomBSplineCurve::getClassTypeId()    ? ( isNamingBoxChecked ?
-                                                                (tr("BSpline") + QString::fromLatin1("(Edge%1#ID%2)").arg(i).arg(i-1)) +
-                                                                (construction?(QString::fromLatin1("-")+tr("Construction")):QString::fromLatin1("")):
-                                                                (QString::fromLatin1("%1-").arg(i)+tr("BSpline")))   :
-        ( isNamingBoxChecked ?
-          (tr("Other") + QString::fromLatin1("(Edge%1#ID%2)").arg(i).arg(i-1)) +
-          (construction?(QString::fromLatin1("-")+tr("Construction")):QString::fromLatin1("")):
-          (QString::fromLatin1("%1-").arg(i)+tr("Other"))),
-        i-1,
-        sketchView->getSketchObject()->getVertexIndexGeoPos(i-1,Sketcher::PointPos::start),
-        sketchView->getSketchObject()->getVertexIndexGeoPos(i-1,Sketcher::PointPos::mid),
-        sketchView->getSketchObject()->getVertexIndexGeoPos(i-1,Sketcher::PointPos::end),
-        type,
-        construction,
-        false));
-
-        setItemVisibility(i-1, mode);
+        Base::Type type = (*it)->getTypeId();
+        bool construction = Sketcher::GeometryFacade::getConstruction(*it);
+        bool internalAligned = Sketcher::GeometryFacade::isInternalAligned(*it);
+        if (internalAligned) 
+            construction = false;
+
+        ElementWidgetIcons::Type icontype = construction ? ElementWidgetIcons::Type::Construction : (internalAligned ? ElementWidgetIcons::Type::Internal : ElementWidgetIcons::Type::Normal);
+
+        ElementItem* itemN = new ElementItem(i - 1,
+            sketchView->getSketchObject()->getVertexIndexGeoPos(i - 1, Sketcher::PointPos::start),
+            sketchView->getSketchObject()->getVertexIndexGeoPos(i - 1, Sketcher::PointPos::mid),
+            sketchView->getSketchObject()->getVertexIndexGeoPos(i - 1, Sketcher::PointPos::end),
+            type, construction, internalAligned, false,
+            ElementWidgetIcons::getIcon(type, Sketcher::PointPos::none, icontype),
+            ElementWidgetIcons::getIcon(type, Sketcher::PointPos::start, icontype),
+            ElementWidgetIcons::getIcon(type, Sketcher::PointPos::end, icontype),
+            ElementWidgetIcons::getIcon(type, Sketcher::PointPos::mid, icontype),

```

Comment:
Can't we just store all the icons once and use pointers to those icons? Though there are a lot of icons (8 x 4 x 4 = ~128 icons).

Reply:
Yes, maybe something as a `std::shared_ptr`.

Reply:
There should be only one copy of each icon (unless QT requires a copy for itself). That is what ElementWidgetIcons is set to achieve. It owns the icons and only provides references to them. It manages their lifetime too.

The original ElementItem constructor was taking references to icons, instead of copies. The new version uses pass-by-value. If we change it to the original, we are just passing the reference. 

It may be the case that we do not even need to store the icons in ElementItem, if ElementItem knows the Type of the geometry, the implementation can get a reference to the the right icons based on this information. That would simplify the code.

I will look into it and come back with a proposal.

Reply:
Maybe I'm missing something, but in the class declaration of `ElementItem`, icons are member variables and not pointers...

Reply:
Yes, if you maintain them, there needed to be references. In my proposal below, they are not needed.

What we are doing here is separating the representation from the model, the item itself, so the view knows how to get the relevant icons for the item.